### PR TITLE
2.5.x Installer

### PR DESCRIPTION
**Pull request by dpsommer**

### CODE REVIEW

**Comment by dpsommer** [comment]
@asantoni the commit reverted in 5e22f929c7833a354f5d40b8d7cc2c6ac52604ec was f4c5cf135364e43ac446629fcba0d63cf79d0af2

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 .*
 *.pyc
+*~
```
Comment (asantoni) [inline .gitignore]: Should be *.*~?

File: README [hunk 2]
```diff
@@ -30,16 +30,51 @@ http://sourcefabric.booktype.pro/airtime-25-for-broadcasters/preparing-the-serve
...
+/usr/share/airtime as your web root (where apache looks for your Airtime files)
+
+There are several options for installation - to see them all, run
+
+    sudo ./install --help
+
+2) Once you've run the installer, navigate to localhost/ (if you have a custom 
+apache configuration, navigate to your Airtime web host) to run an interactive 
+setup.
+
```
Comment (asantoni) [inline README:66]: I just committed some minor clarification of this to your branch

File: airtime_mvc/application/models/Show.php [hunk 3]
```diff
@@ -1273,7 +1272,6 @@ public static function getPrevCurrentNextOld($utcNow)
                             "starts"                  => $rows[$i-1]['starts'],
                             "ends"                    => $rows[$i-1]['ends'],
                             "record"                  => $rows[$i-1]['record'],
-                            "image_path"             => $rows[$i-1]['image_path'],
```
Comment (asantoni) [inline airtime_mvc/application/models/Show.php:1287]: Why is this removed?

**Comment by asantoni** [comment]
Is there a reason why the airtime-setup folder is in build?

File: airtime_mvc/build/airtime-setup/load.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        "postgres" => checkDatabaseDependencies()
+    );
+}
+
+/**
+ * Check that the Zend framework libraries are installed
+ *
+ * @return boolean true if Zend exists in /usr/share/php
+ */
+function checkMvcDependencies() {
```
Comment (asantoni) [inline airtime_mvc/build/airtime-setup/load.php]: Why not name this checkZendDependencies?

File: airtime_mvc/build/airtime-setup/load.php [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    return isset($conn);
+}
+
+/**
+ * Check if airtime-media-monitor is currently running
+ * 
+ * @return boolean true if airtime-media-monitor is running
+ */
+function checkMediaMonitorService() {
+    exec("service airtime-media-monitor status | grep start", $out, $status);
```
Comment (asantoni) [inline airtime_mvc/build/airtime-setup/load.php]: It'd be nice if these exec calls were abstracted out into a platform layer (since they're Linux-only)...

File: airtime_mvc/public/css/setup/setup-config.css [hunk 6]
```diff
@@ -0,0 +1,219 @@
+html {
+    background: -webkit-linear-gradient(midnightblue, black); /* For Safari 5.1 to 6.0 */
+    background: -o-linear-gradient(midnightblue, black); /* For Opera 11.1 to 12.0 */
+    background: -moz-linear-gradient(midnightblue, black); /* For Firefox 3.6 to 15 */
+    background: linear-gradient(midnightblue, black); /* Standard syntax */
```
Comment (asantoni) [inline airtime_mvc/public/css/setup/setup-config.css]: At some point I'm going to tone down the blue... :)

File: airtime_mvc/public/setup/media-setup.php [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+require_once(dirname(dirname( __DIR__)) . "/application/models/airtime/map/CcMusicDirsTableMap.php");
+require_once(dirname(dirname( __DIR__)) . "/application/models/airtime/om/BaseCcMusicDirsQuery.php");
+require_once(dirname(dirname( __DIR__)) . "/application/models/airtime/CcMusicDirsQuery.php");
+require_once(dirname(dirname( __DIR__)) . "/application/models/airtime/om/BaseCcMusicDirs.php");
+require_once(dirname(dirname( __DIR__)) . "/application/models/airtime/CcMusicDirs.php");
+require_once(dirname(dirname( __DIR__)) . "/application/models/airtime/om/BaseCcMusicDirsPeer.php");
+require_once(dirname(dirname( __DIR__)) . "/application/models/airtime/CcMusicDirsPeer.php");
+
+/**
+ * User: sourcefabric
```
Comment (asantoni) [inline airtime_mvc/public/setup/media-setup.php]: user ?

File: airtime_mvc/public/setup/media-setup.php [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+        $this->runMusicDirsQuery($con);
+    }
+
+    function runMusicDirsQuery($con) {
+        try {
+            if ($this->checkMusicDirectoryExists($con)) {
+                $dir = CcMusicDirsQuery::create()->findPk(1, $con);
+            } else {
+                $dir = new CcMusicDirs();
```
Comment (asantoni) [inline airtime_mvc/public/setup/media-setup.php]: I wish this nitty gritty was actually inside the rest of Airtime....

File: install [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+verbose "...Done"
+
+verbose "\n * Creating /run/airtime..."
+mkdir -p /run/airtime
+chmod 755 /run/airtime
+chown -R ${web_user}:${web_user} /run/airtime
+verbose "...Done"
+
+verbose "\n * Copying logging files..."
+cp -R ${AIRTIMEROOT}/python_apps/std_err_override /usr/lib/airtime/
```
Comment (asantoni) [inline install]: Why is this necessary?

File: install [hunk 10]
```diff
@@ -0,0 +1,593 @@
...
+verbose "...Done"
+
+verbose "\n * Copying API client files..."
+cp -R ${AIRTIMEROOT}/python_apps/api_clients /usr/lib/airtime/
+verbose "...Done"
+
```
Comment (asantoni) [inline install:457]: Same, why is this not installed with setuptools?

File: install [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+verbose "\n * Installing media-monitor..."
+python ${AIRTIMEROOT}/python_apps/media-monitor/setup.py install
+verbose "...Done"
+
+verbose "\n * Copying pypo files..."
+mkdir -p /usr/lib/airtime/pypo
+cp -R ${AIRTIMEROOT}/python_apps/pypo/pypo /usr/lib/airtime/pypo/bin/
+verbose "...Done"
```
Comment (asantoni) [inline install:464]: Ditto for the above manual cps - why do we need to do that? setuptools should be doing whatever is necessary to install those files. 

Consider what happens if you're not on a debian-based Linux distro. You should be able to do sudo python setup.py install for media-monitor and pypo, and point an Apache site to your airtime directory, and it should work.

File: install [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+if [ ! -d /var/log/airtime ]; then
+    loud "\n-----------------------------------------------------"
+    loud "              * Installing Log Files *               "
+    loud "-----------------------------------------------------"
+
+    verbose "\n * Creating /var/log/airtime..."
+    mkdir -p /var/log/airtime
+    mkdir -p /var/log/airtime/media-monitor
+    mkdir -p /var/log/airtime/pypo
+    mkdir -p /var/log/airtime/pypo-liquidsoap
```
Comment (asantoni) [inline install]: Each component's individual installer should create any directories it needs. 
eg. The setuptools-based installer for pypo should create /var/log/airtime/pypo and /var/log/airtime/pypo-liquidsoap.

File: install [hunk 13]
```diff
@@ -0,0 +1,593 @@
...
+    loudCmd "apt-get -y --force-yes install zendframework"
+else
+    loudCmd "apt-get -y --force-yes install libzend-framework-php"
+fi
+
+# PHP Config File for Apache
+if [ ! -f "/etc/php5/apache2/conf.d/airtime.ini" ]; then
+    verbose "\n * Creating Airtime PHP config for Apache..."
+    cp ${SCRIPT_DIR}/installer/php/airtime.ini /etc/php5/apache2/conf.d/airtime.ini
+else
```
Comment (asantoni) [inline install:509]: Do we still need this custom airtime.ini for Apache?

File: installer/lib/requirements-debian-wheezy.apt [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+lsof
+mp3gain
+vorbisgain
+flac
+vorbis-tools
+pwgen
+libfaad2
+php-apc
+
+libmp3lame-dev
```
Comment (asantoni) [inline installer/lib/requirements-debian-wheezy.apt]: This -dev package is not needed.

File: installer/lib/requirements-debian-wheezy.apt [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+python
+python-virtualenv
+
+libsoundtouch-ocaml
+libtaglib-ocaml
+libao-ocaml
+libmad-ocaml
+ecasound
+libesd0
```
Comment (asantoni) [inline installer/lib/requirements-debian-wheezy.apt]: Can you check if libesd0 is still needed? It's super archaic...

File: installer/lib/requirements-ubuntu-saucy.apt [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+python
+python-virtualenv
+
+libsoundtouch-ocaml
+libtaglib-ocaml
+libao-ocaml
+libmad-ocaml
+ecasound
+libesd0
```
Comment (asantoni) [inline installer/lib/requirements-ubuntu-saucy.apt]: I'm almost certain you don't need libesd0 here

File: installer/lib/requirements-ubuntu-saucy.apt [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+lsof
+mp3gain
+vorbisgain
+flac
+vorbis-tools
+pwgen
+libfaad2
+php-apc
+
+libmp3lame-dev
```
Comment (asantoni) [inline installer/lib/requirements-ubuntu-saucy.apt]: remove libmp3lame-dev

File: installer/lib/requirements-ubuntu-trusty.apt [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+python
+python-virtualenv
+
+libsoundtouch-ocaml
+libtaglib-ocaml
+libao-ocaml
+libmad-ocaml
+ecasound
+libesd0
```
Comment (asantoni) [inline installer/lib/requirements-ubuntu-trusty.apt]: libesd0 again

File: installer/lib/requirements-ubuntu-trusty.apt [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+lsof
+mp3gain
+vorbisgain
+flac
+vorbis-tools
+pwgen
+libfaad2
+php-apc
+
+libmp3lame-dev
```
Comment (asantoni) [inline installer/lib/requirements-ubuntu-trusty.apt]: libmp3lame-dev again

File: python_apps/media-monitor/bin/airtime-media-monitor [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+exec 2>&1
+
+cd /usr/lib/airtime/media-monitor/
+
+set +e
+cat /etc/default/locale | grep -i "LANG=.*UTF-\?8"
+set -e
+if [ "$?" != "0" ]; then
+    echo "non UTF-8 default locale found in /etc/default/locale." > /var/log/airtime/media-monitor/error.log
+    exit 1
```
Comment (asantoni) [inline python_apps/media-monitor/bin/airtime-media-monitor]: Why is this check needed?

File: python_apps/media-monitor/bin/airtime-media-monitor [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+fi
+
+export PYTHONPATH=/usr/lib/airtime/:/usr/lib/airtime/media-monitor/airtimefilemonitor/:\
```
Comment (asantoni) [inline python_apps/media-monitor/bin/airtime-media-monitor]: Can we refactor media-monitor to not need this custom python path? It needlessly complicates things.

File: python_apps/media-monitor/bin/airtime-media-monitor [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+/usr/lib/airtime/media-monitor/mm2/:/usr/lib/airtime/std_err_override/:/usr/lib/airtime/api_clients:$PYTHONPATH
+export LC_ALL=`cat /etc/default/locale | grep "LANG=" | cut -d= -f2 | tr -d "\n\""`
+
```
Comment (asantoni) [inline python_apps/media-monitor/bin/airtime-media-monitor]: More weird pypo stuff.

File: python_apps/media-monitor/bin/airtime-media-monitor [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+exec python /usr/lib/airtime/media-monitor/media_monitor.py > /var/log/airtime/media-monitor/py-interpreter.log 2>&1
```
Comment (asantoni) [inline python_apps/media-monitor/bin/airtime-media-monitor]: Logging without logrotate is not great...

File: python_apps/media-monitor/install/airtime-media-monitor.conf [hunk 24]
```diff
@@ -0,0 +1,15 @@
+description "Airtime Media Monitor"
+author "help@sourcefabric.org"
+
+start on runlevel [2345]
+stop on runlevel [!2345]
+
+respawn
+
+setuid WEB_USER
```
Comment (asantoni) [inline python_apps/media-monitor/install/airtime-media-monitor.conf]: If all of these files are being templated with WEB_USER, should we rename them to blah.conf.template instead, to make it clearer?

File: python_apps/media-monitor/setup.py [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+          'kombu',
+          'mutagen',
+          'poster',
+          'PyDispatcher',
+          'pyinotify',
+          'pytz',
+          'wsgiref'
+      ],
+      zip_safe=False,
+      data_files=data_files)
```
Comment (asantoni) [inline python_apps/media-monitor/setup.py:60]: As mentioned earlier, we should create any needed log folders here, or media-monitor should be able to create /var/log/airtime/media-monitor itself. (I think the permissions are fixed now so that this is possible... ?)

File: python_apps/pypo/bin/airtime-liquidsoap [hunk 26]
```diff
@@ -22,27 +22,17 @@ do
...
+    ls_path="/usr/bin/airtime-liquidsoap --verbose -f"
 fi
-ls_param="/usr/lib/airtime/pypo/bin/liquidsoap_scripts/ls_script.liq"
-
-export PYTHONPATH=${api_client_path}
 
-SCRIPT=`readlink -f $0`
-# Absolute directory this script is in
-SCRIPTPATH=`dirname $SCRIPT`
+export PYTHONPATH=/usr/lib/airtime/:/usr/lib/airtime/pypo/bin/:/usr/lib/airtime/std_err_override/:/usr/lib/airtime/api_clients:$PYTHONPATH
```
Comment (asantoni) [inline python_apps/pypo/bin/airtime-liquidsoap]: More PYTHONPATH sketchiness

File: python_apps/pypo/bin/airtime-playout [hunk 27]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+set +e
+cat /etc/default/locale | grep -i "LANG=.*UTF-\?8" > /dev/null
+set -e
+if [ "$?" != "0" ]; then
+    echo "non UTF-8 default locale found in /etc/default/locale." > /var/log/airtime/pypo/error.log
+    exit 1
+fi
```
Comment (asantoni) [inline python_apps/pypo/bin/airtime-playout]: Ditto to the everything I said about the media-monitor script

File: python_apps/pypo/pypo/liquidsoap_scripts/liquidsoap_prepare_terminate.py [hunk 28]
```diff
@@ -14,6 +14,6 @@
     tn.read_all()
 
 except Exception, e:
-    print('Error loading config file: %s', e)
+    print 'Error loading config file: %s' % e
```
Comment (asantoni) [inline python_apps/pypo/pypo/liquidsoap_scripts/liquidsoap_prepare_terminate.py]: You replaced a line that's forward compatible with Python 3-compatible with a line that's only compatible with python 2...

File: python_apps/pypo/setup.py [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+          'mutagen',
+          'poster',
+          'PyDispatcher',
+          'pyinotify',
+          'pytz',
+          'wsgiref'
+      ],
+      zip_safe=False,
+      data_files=data_files)
+
```
Comment (asantoni) [inline python_apps/pypo/setup.py:65]: Create the log dirs here?

File: uninstall [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+rm -rf /usr/lib/airtime/
+
+rm -rf /usr/share/airtime
+
+rm -f /etc/init/airtime*
+rm -f /usr/bin/airtime*
+
+rm -f /etc/apache2/sites-enabled/airtime*
+rm -f /etc/apache2/sites-available/airtime*
+
```
Comment (asantoni) [inline uninstall:95]: We should DEFINITELY prompt before dropping the database...

File: uninstall [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+rm -rf /var/log/airtime/
+rm -rf /usr/lib/airtime/
+
+rm -rf /usr/share/airtime
+
+rm -f /etc/init/airtime*
+rm -f /usr/bin/airtime*
+
+rm -f /etc/apache2/sites-enabled/airtime*
+rm -f /etc/apache2/sites-available/airtime*
```
Comment (asantoni) [inline uninstall]: These rm -rf calls should all pull from the FILES array, not have a second hardcoded set of dirs/files there...

File: utils/soundcloud-uploader.php [hunk 32]
```diff
@@ -1,52 +1,31 @@
...
-set_include_path($CC_CONFIG['phpDir']."/application/models" . PATH_SEPARATOR . get_include_path());
-require_once($CC_CONFIG['phpDir']."/library/propel/runtime/lib/Propel.php");
-Propel::init($CC_CONFIG['phpDir']."/application/configs/airtime-conf.php");
+set_include_path($webRoot.'/library' . PATH_SEPARATOR . get_include_path());
+require_once($webRoot.'/application/models/Soundcloud.php');
 
+set_include_path($webRoot."/application/models" . PATH_SEPARATOR . get_include_path());
 require_once 'propel/runtime/lib/Propel.php';
-Propel::init($CC_CONFIG['phpDir']."/application/configs/airtime-conf-production.php");
+Propel::init($webRoot."/application/configs/airtime-conf-production.php");
```
Comment (asantoni) [inline utils/soundcloud-uploader.php:28]: We have to make sure to test SoundCloud upload, because I guarantee nobody's tested this in the last year...

File: airtime_mvc/application/models/Show.php (hunk 3 above)
Comment (dpsommer) [inline airtime_mvc/application/models/Show.php:1287]: Need to fix this still; once the upgrade is introduced (and, more importantly, tested) I'll go and grab the relevant code from upstream to reintroduce the showart/instance description

File: airtime_mvc/build/airtime-setup/load.php (hunk 4 above)
Comment (dpsommer) [inline airtime_mvc/build/airtime-setup/load.php]: Fixed

File: airtime_mvc/public/css/setup/setup-config.css (hunk 6 above)
Comment (dpsommer) [inline airtime_mvc/public/css/setup/setup-config.css]: BLASPHEMY

File: airtime_mvc/public/setup/media-setup.php (hunk 7 above)
Comment (dpsommer) [inline airtime_mvc/public/setup/media-setup.php]: Fixed

File: install (hunk 9 above)
Comment (dpsommer) [inline install]: This needs to be addressed alongside an overhaul of how pypo and media-monitor (possibly also api_client) work

File: install (hunk 12 above)
Comment (dpsommer) [inline install]: Fixed

File: install (hunk 13 above)
Comment (dpsommer) [inline install:509]: Leaving this in as there may be unexpected corner cases that need it

File: installer/lib/requirements-debian-wheezy.apt (hunk 14 above)
Comment (dpsommer) [inline installer/lib/requirements-debian-wheezy.apt]: Fixed

File: installer/lib/requirements-debian-wheezy.apt (hunk 15 above)
Comment (dpsommer) [inline installer/lib/requirements-debian-wheezy.apt]: Fixed

File: installer/lib/requirements-ubuntu-saucy.apt (hunk 16 above)
Comment (dpsommer) [inline installer/lib/requirements-ubuntu-saucy.apt]: Fixed

File: installer/lib/requirements-ubuntu-saucy.apt (hunk 17 above)
Comment (dpsommer) [inline installer/lib/requirements-ubuntu-saucy.apt]: Fixed

File: installer/lib/requirements-ubuntu-trusty.apt (hunk 18 above)
Comment (dpsommer) [inline installer/lib/requirements-ubuntu-trusty.apt]: Fixed

File: installer/lib/requirements-ubuntu-trusty.apt (hunk 19 above)
Comment (dpsommer) [inline installer/lib/requirements-ubuntu-trusty.apt]: Fixed

File: python_apps/media-monitor/bin/airtime-media-monitor (hunk 20 above)
Comment (dpsommer) [inline python_apps/media-monitor/bin/airtime-media-monitor]: Leaving this in until we can take a closer look at it when overhauling pypo/media-monitor

File: python_apps/media-monitor/install/airtime-media-monitor.conf (hunk 24 above)
Comment (dpsommer) [inline python_apps/media-monitor/install/airtime-media-monitor.conf]: Fixed; changed all upstart conf files to .templates

File: python_apps/pypo/pypo/liquidsoap_scripts/liquidsoap_prepare_terminate.py (hunk 28 above)
Comment (dpsommer) [inline python_apps/pypo/pypo/liquidsoap_scripts/liquidsoap_prepare_terminate.py]: Fixed

File: uninstall (hunk 31 above)
Comment (dpsommer) [inline uninstall]: Fixed

File: uninstall (hunk 30 above)
Comment (dpsommer) [inline uninstall:95]: Fixed
